.github/workflows/build-and-test.yml

```diff
@@ -1,91 +1,59 @@
 
 on:
-  push:
-    branches: [ master ]
   pull_request:
-    branches: [ master ]
+  push:
+  schedule:
+    # Prime the caches every Monday
+    - cron: 0 1 * * MON
 
 jobs:
   build-and-test:
     name: Build and Test
 
-    runs-on: ${{ matrix.os }}
-
     strategy:
       fail-fast: false
       matrix:
         os:
           - macos-latest
           - ubuntu-latest
           - windows-latest
+        ocaml-compiler:
+          - 4.12.x
 
-        ocaml-version:
-          - 4.12.0
+    runs-on: ${{ matrix.os }}
 
     steps:
       - name: Checkout code
         uses: actions/checkout@v2
         with:
           submodules: true
 
-      - name: Retrieve opam cache
-        uses: actions/cache@v2
-        if: runner.os != 'Windows'
-        id: cache-opam
-        with:
-          path: ~/.opam
-          key: v1-${{ runner.os }}-opam-${{ matrix.ocaml-version }}-${{ hashFiles('*.opam.locked') }}
-          restore-keys: |
-            v1-${{ runner.os }}-opam-${{ matrix.ocaml-version }}-
-
       - name: Use latest LTS Node.js release
-        uses: actions/setup-node@v1
+        uses: actions/setup-node@v2
         with:
           node-version: 14.x
 
-      - name: Use OCaml ${{ matrix.ocaml-version }}
-        uses: avsm/setup-ocaml@v1
-        with:
-          ocaml-version: ${{ matrix.ocaml-version }}
-
-      - name: Print OCaml version
-        run: opam exec -- ocaml --version
-
-      - name: Pin local packages
-        run: |
-          opam pin jsonrpc.dev . --no-action
-          opam pin lsp.dev . --no-action
-          opam pin ocaml-lsp-server.dev . --no-action
-
-      - name: Install OCaml secondary compiler
-        if: steps.cache-opam.outputs.cache-hit != 'true' && runner.os != 'Windows' && (matrix.ocaml-version == '4.06.1')
-        run: opam install ocamlfind-secondary
+      - run: yarn --frozen-lockfile
+        working-directory: ocaml-lsp-server/test/e2e
 
-      - name: Install Opam dependencies
-        if: steps.cache-opam.outputs.cache-hit != 'true'
-        run: |
-          opam install . --with-doc --with-test --locked --ignore-constraints-on=ocaml
-          opam install ocamlformat.0.17.0 --skip-updates
+      - name: Use OCaml ${{ matrix.ocaml-compiler }}
+        uses: avsm/setup-ocaml@v2
+        with:
+          ocaml-compiler: ${{ matrix.ocaml-compiler }}
+          dune-cache: ${{ matrix.os != 'macos-latest' }}
+          opam-depext-flags: --with-test
 
-      - name: Fix broken opam state
-        if: steps.cache-opam.outputs.cache-hit == 'true'
-        run: opam upgrade --fixup
+      - run: opam depext ocamlformat=$(cat .ocamlformat | grep version | cut -d '=' -f 2) --install
 
-      - name: Install Node dependencies
-        working-directory: ocaml-lsp-server/test/e2e
-        run: |
-          yarn --frozen-lockfile
+      - run: opam install . --with-test
 
-      - name: Build
-        run: opam exec -- make all
+      - run: opam exec -- make all
 
-      - name: Check formatting
```

Comment:
Which action is doing the formatting check now?

Reply:
It does at the last, but auto promote is currently enabled, so it's actually being formatted instead of checked. Somehow, the check-fmt task seems to have been removed at some point.

Reply:
Even if auto promote is enabled, the error code should still indicate failure if there was a diff I think. In any case, it's better to verify that formatting errors fail the CI

Reply:
If "always" is enabled, the job is finally marked as failed after running to the last step, so the current configuration seems quite reasonable.